plugins/batterystat.koplugin/main.lua

```diff
@@ -0,0 +1,225 @@
+    settings = LuaSettings:open(DataStorage:getSettingsDir() .. "/batterstat.lua"),
+    dump_file = util.realpath(DataStorage:getDataDir()) .. "/battery_stat.txt",
+}
+
+function BatteryStat:init()
+    self.charging = Usage:new(self.settings.charging)
+    self.decharging = Usage:new(self.settings.decharging)
+    self.awake = Usage:new(self.settings.awake)
+    self.sleeping = Usage:new(self.settings.sleeping)
+
+    -- Note: these fields are not the "real" timestamp and battery usage, but
+    -- the unaccumulated values.
+    self.charging_state = State:new(self.settings.charging_state)
+    self.awake_state = State:new(self.settings.awake_state)
+    -- Whether the device was suspending before current timestamp.
+    self.was_suspending = false
+    -- Whether the device was charging before current timestamp.
+    self.was_charging = PowerD:isCharging()
+
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function BatteryStat:onFlushSettings()
+    self.settings:replace({
+        charging = self.charging,
+        decharging = self.decharging,
+        awake = self.awake,
+        sleeping = self.sleeping,
+        charging_state = self.charging_state,
+        awake_state = self.awake_state,
+    })
+    self.settings:flush()
+end
+
+function BatteryStat:accumulate()
+    if self.was_suspending then
+        -- Suspending to awake.
+        self.sleeping:append(self.awake_state)
+    else
+        -- Awake to suspending, time between self.awake_state and now should belong to awake.
+        self.awake:append(self.awake_state)
+    end
+    if self.was_charging then
+        -- Decharging to charging.
+        self.charging:append(self.charging_state)
+    else
+        self.decharging:append(self.charging_state)
+    end
+    self.awake_state = State:new()
+    self.charging_state = State:new()
+end
+
+function BatteryStat:onSuspend()
+    self.was_suspending = false
+    self:accumulate()
+end
+
+function BatteryStat:onResume()
+    self.was_suspending = true
+    self:accumulate()
+end
+
+function BatteryStat:onCharging()
+    self.was_charging = false
+    self:dumpToText()
+    self.charging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onNotCharging()
+    self.was_charging = true
+    self:dumpToText()
+    self.decharging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:dumpToText()
+    local kv_pairs = self:dump()
+    local content = T(_("Dump at %1"), os.date("%c"))
+    for _, pair in ipairs(kv_pairs) do
+        content = content .. "\n" .. pair[1]
+        if pair[2] ~= nil and pair[2] ~= "" then
+            content = content .. "\t" .. pair[2]
+        end
+    end
+    content = content .. "\n-=-=-=-=-=-\n"
+    local file = io.open(self.dump_file, "a")
+    if file then
+        file:write(content)
+        file:close()
+    else
+        logger.warn(content)
```

Comment:
can we add more context to this warning log? For example, `logger.warn("failed to open batterystats log file", content)`